### PR TITLE
Dispatcher and structured kernel lab

### PR DESCRIPTION
**Pull request by arunppsg**
I am working on the [dispatcher and structured kernels lab](https://github.com/pytorch/pytorch/wiki/Dispatcher---Structured-Kernels-Lab) in this pull request.

### CODE REVIEW

**Comment by arunppsg** [comment]
`RegisterCPU.cpp` has the following defintions
```cpp
struct structured_mul_out_functional final : public at::native::structured_mul_out {
struct structured_mul_out_out final : public at::native::structured_mul_out {
```
If I am not wrong, these are the lines which call into the CPU kernel. Since entries for `mul` is modified in `native_functions.yaml` and codegen will not generate these lines, I am not sure where to add these definitions.

**Comment by bdhirsh** [comment]
> If I am not wrong, these are the lines which call into the CPU kernel. Since entries for mul is modified in native_functions.yaml and codegen will not generate these lines, I am not sure where to add these definitions.

That's right - those 3 structs (and basically everything inside of `RegisterCPU.cpp`) include some of the boilerplate required to write the cpu kernel. Two things:

(1) Since the codegen is no longer generating it, you can write it manually.

(2) There are a few ways you can choose to decide what API's call into what - you need to make sure that `at::mul`, `at::Tensor::mul` and `at::native::mul` are all implemented. The easiest way is probably to have both `at::mul` and `at::Tensor::mul` call into `at::native::mul`, and implement `at::native::mul` similarly to how the mul kernel is implemented in `RegisterCPU.cpp` (basically just by copying the kernel that was originally generated there from the codegen).

Happy to answer any more specific questions!

**Comment by arunppsg** [comment]
If I understand correctly, I could either define at::Tensor api in `TensorMethods.cpp` or define at::native api in `BinaryOps.cpp`? In the latter case, we have to make at::Tensor::mul call into at::native::mul

also, what does `TORCH_API` macro do?

**Comment by arunppsg** [comment]
I am doing some mistake which I am not able to figure out - maybe a missing header because when I build it, I get the following error traceback:
```
In file included from /Users/arun/proj/torch/pytorch/aten/src/ATen/LegacyBatchingRegistrations.cpp:4:
In file included from /Users/arun/proj/torch/pytorch/aten/src/ATen/LegacyBatchedFallback.h:2:
In file included from /Users/arun/proj/torch/pytorch/aten/src/ATen/ATen.h:13:
In file included from /Users/arun/proj/torch/pytorch/build/aten/src/ATen/Functions.h:76:
/Users/arun/proj/torch/pytorch/build/aten/src/ATen/NativeFunctions.h:1281:56: error: unknown class name 'structured_mul_Tensor'; did you mean 'structured_sub_Tensor'?
```

File: aten/src/ATen/templates/NativeFunctions.h [hunk 1]
```diff
@@ -31,3 +31,18 @@
 ${NativeFunctions_includes}
 
 ${NativeFunctions_declarations}
+
+namespace at {
+namespace native {
+
+// Added from torch/include/ATen/ops/mul_native.h
+struct TORCH_API structured_mul_out : public at::meta::structured_mul_Tensor {
```
Comment (bdhirsh) [inline aten/src/ATen/templates/NativeFunctions.h:39]: If you take a look at the built `NativeMetaFunctions.h` file, you'll see that we don't directly `#include <ATen/NativeMetaFunctions.h>` in it. So if you're going to defined the `at::meta::structured_mul_Tensor` struct in that file, you'll need to manually include that in this file too.

**Pull request closed without merging**